File: traverse_tree/py/TestTraversal.py
from TreeNode import *
from collections import deque
import random
import unittest

from StackTraversal import StackTraversal
from MorrisTraversal import MorrisTraversal
from RecursiveTraversal import RecursiveTraversal

class TestTraversal(unittest.TestCase):
    def __random_tree__(self, size):
        pool = set( [str(i) for i in range(size)] )
        cnt = 1
        root = TreeNode( pool.pop() )
        queue = deque([root])
        while queue and cnt < size:
            node = queue.popleft()
            if pool and (random.random() < 0.7 or not queue):
                node.left = TreeNode( pool.pop() )
                queue.append( node.left )
            if pool and (random.random() < 0.7 or not queue):
                node.right = TreeNode( pool.pop() )
                queue.append( node.right )
            cnt += 1
        return root

    def __testTraversalMethod__(self, impl, size = 100, case = None):
        root = case or self.__random_tree__(size)
        tag = ["preorder", "inorder", "postorder"]

        ans = [0 for i in range(3)]
        ans[0] = RecursiveTraversal.preorderTraversal(root)
        ans[1] = RecursiveTraversal.inorderTraversal(root)
        ans[2] = RecursiveTraversal.postorderTraversal(root)

        ret = [0 for i in range(3)]
        ret[0] = impl.preorderTraversal(root)
        ret[1] = impl.inorderTraversal(root)
        ret[2] = impl.postorderTraversal(root)

        self.assertEqual(ans[0], ret[0])
        self.assertEqual(ans[1], ret[1])
        self.assertEqual(ans[2], ret[2])

    def __testOtherMethods__(self, size = 100):
        root = self.__random_tree__(size)
        tag = ["preorder", "inorder", "postorder"]

        ans = [0 for i in range(3)]
        ans[0] = RecursiveTraversal.preorderTraversal(root)
        ans[1] = RecursiveTraversal.inorderTraversal(root)
        ans[2] = RecursiveTraversal.postorderTraversal(root)

        ret = [0 for i in range(3)]
        ret[0] = StackTraversal.dfsPreorderTraversal(root)
        ret[1] = StackTraversal.inorderTraversal(root)
        ret[2] = StackTraversal.psudoPostorderTraversal(root)

        self.assertEqual(ans[0], ret[0])
        self.assertEqual(ans[1], ret[1])
        self.assertEqual(ans[2], ret[2])

    def testStackTraversal(self):
        for i in range(100):
            self.__testTraversalMethod__(StackTraversal, 200)

    def testMorrisTraversal(self):
        for i in range(100):
            self.__testTraversalMethod__(MorrisTraversal, 200)

    def testOtherMethods(self):
        for i in range(100):
            self.__testOtherMethods__()

if __name__ == '__main__':
    unittest.main()

File: traverse_tree/java/TestTraversal.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.ArrayDeque;
import java.util.Random;
import java.lang.Integer;

public class TestTraversal
{
	private static Random myrand = new Random();
	private static int pickAndTake(List<Integer> pool, int size)
	{
		int randPick = myrand.nextInt(size);
		int result = pool.get(randPick);
		pool.set(randPick, pool.get(size - 1));
		return result;
	}
	public static TreeNode generateRandomTree(int size)
	{
		List<Integer> pool = new ArrayList<Integer>();
		for(int i  = 0; i < size; i++)
			pool.add(i);
		Collections.shuffle(pool);

		TreeNode root = new TreeNode(pickAndTake(pool, size--));
		ArrayDeque<TreeNode> queue = new ArrayDeque<TreeNode>();
		queue.addLast(root);
		for(int i  = 1; i < size - 1; i++)
		{
			TreeNode node = queue.pollFirst();
			if(myrand.nextFloat() < 0.7)
			{
				node.left = new TreeNode(pickAndTake(pool, size--));
				queue.addLast(node.left);
				i++;
			}
			if(myrand.nextFloat() < 0.7 || queue.peekFirst() == null)
			{
				node.right = new TreeNode(pickAndTake(pool, size--));
				queue.addLast(node.right);
				i++;
			}
		}
		return root;
	}
	public static boolean testTraversalMethod(TraversalMethods impl, int size)
	{
		TreeNode root = generateRandomTree(size);
		String tag[] = { "preorder", "inorder", "postorder" };
		RecursiveTraversal rcur = new RecursiveTraversal();

		List<List<Integer>> ans = new ArrayList<List<Integer>>();
		ans.add( rcur.preorderTraversal(root) );
		ans.add( rcur.inorderTraversal(root) );
		ans.add( rcur.postorderTraversal(root) );

		ArrayList<List<Integer>> ret = new ArrayList<List<Integer>>();
		ret.add( impl.preorderTraversal(root) );
		ret.add( impl.inorderTraversal(root) );
		ret.add( impl.postorderTraversal(root) );

		for(int i = 0; i < 3; i++)
		{
			if(ret.get(i).size() != ans.get(i).size())
			{
				System.out.println(tag[i] + " answer length error");
				return false;
			}
			for(int j = 0; j < ans.get(i).size(); j++)
				if(ret.get(i).get(j).compareTo(ans.get(i).get(j)) != 0)
				{
					System.out.println(tag[i] + " result error");
					return false;
				}
		}
		return true;
	}
	public static void main(String[] argv)
	{
		TraversalMethods impls[] = {
			new StackTraversal(), new MorrisTraversal() };
		for(TraversalMethods impl: impls)
		{
			int cases = 500;
			for(int i = 0; i < cases; i++)
				if(!testTraversalMethod(impl, 200))
				{
					System.out.println("WA: " + impl);
					break;
				}
		}
	}
}
